File: app/routes/inventory_adjustment.py
from flask import Blueprint, render_template, request, jsonify
import logging
import json
import requests
import uuid
from datetime import datetime
import time
from config import config

# 设置当前模块的日志级别为DEBUG
existing_logger = logging.getLogger()
existing_logger.setLevel(logging.DEBUG)
logger = logging.getLogger(__name__)
logger.setLevel(logging.DEBUG)

# ==================== 蓝图定义 ====================
inventory_adjustment_bp = Blueprint('inventory_adjustment', __name__, url_prefix='/inventory_adjustment')

# ==================== 路由函数 ====================
# 库存调整页面（GET请求）
@inventory_adjustment_bp.route('/')
def index():
    """库存调整页面"""
    return render_template('inventory_adjustment.html')

# 库存调整接口（POST请求）
@inventory_adjustment_bp.route('/submit', methods=['POST'])
def submit():
    """库存调整提交接口"""
    try:
        # 记录请求开始
        logger.info(f"收到库存调整请求，请求方法: POST, 客户端IP: {request.remote_addr}")
        
        # 1. 获取用户输入
        sku_code = request.form.get('skuCode')
        quantity = int(request.form.get('quantity', 0))
        warehouse_code = request.form.get('warehouseCode')
        api_env = request.form.get('apiEnv', 'test')  # 默认测试环境
        
        # 详细记录用户输入
        logger.info(f"用户输入参数: sku_code={sku_code}, quantity={quantity}, warehouse_code={warehouse_code}, api_env={api_env}")
        
        # 2. 验证必填字段
        required_fields = {
            'skuCode': sku_code,
            'quantity': quantity,
            'warehouseCode': warehouse_code
        }
        
        missing_fields = [field for field, value in required_fields.items() if not value]
        if missing_fields:
            logger.warning(f'请求验证失败: 缺少必填字段 {", ".join(missing_fields)}')
            return jsonify({
                'status': 'error',
                'message': f'缺少必填字段: {", ".join(missing_fields)}'
            }), 400
        
        # 3. 选择API地址
        api_url = (
            'https://gateway-test.babycare.com/open/srm/purchase/saveArrival'
            if api_env == 'test' else
            'https://gateway-uat.babycare.com/open/srm/purchase/saveArrival'
        )
        
        logger.info(f"根据环境选择API地址: {api_url}")
        
        # 4. 生成动态参数
        current_time = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        # 使用更可靠的方式生成唯一business_no
        business_no = str(uuid.uuid4()).replace('-', '')[:13]  # 生成13位唯一编号
        # 确保source_code唯一性，添加时间戳后几位
        time_suffix = str(int(time.time() * 1000))[-5:]
        source_code = f"{business_no}{time_suffix}"  # 生成唯一的sourceCode
        
        logger.debug(f"生成动态参数: business_no={business_no}, source_code={source_code}, current_time={current_time}")
        
        # 5. 构建请求参数
        request_data = {
            "appointmentNo": "",
            "businessNo": business_no,
            "checkMethod": "20",
            "detailList": [
                {
                    "actualArrivalNumber": quantity,
                    "batchCode": "1",
                    "lineNo": "1",
                    "planArrivalNumber": quantity,
                    "productDate": "2025-07-01",
                    "sampleQuality": 7,
                    "skuCode": sku_code,
                    "sourceCode": source_code,
                    "volume": 1,
                    "weight": 1
                }
            ],
            "forecastArrivalTime": current_time,
            "forecastDeliveryTime": current_time,
            "isCallCar": 0,
            "orderCreator": "LY",
            "remark": "LY",
            "sourceType": "WWJG",
            "supplierCode": "CO00049541",
            "totalVolume": 1,
            "totalWeight": 1,
            "warehouseCode": warehouse_code,
            "warehouseOutCode": warehouse_code
        }
        
        # 6. 记录请求日志
        logger.info(f"提交库存调整请求: URL={api_url}, 参数={json.dumps(request_data)}")
        
        # 7. 调用外部API（实际环境中会取消注释）
        try:
            # 实际API调用
            logger.info(f"开始调用外部API: {api_url}")
            headers = {'Content-Type': 'application/json'}
            response = requests.post(api_url, json=request_data, headers=headers, timeout=30)
            logger.info(f"外部API响应状态码: {response.status_code}")
            response.raise_for_status()
            result = response.json()
            logger.info(f"外部API调用成功，返回数据: {json.dumps(result)[:500]}...")
        except Exception as api_error:
            logger.error(f"外部API调用失败: {str(api_error)}")
            raise
        
        # 8. 构建并返回响应
        response_data = {
            'status': 'success',
            'message': '库存调整请求已成功提交',
            'data': result,
            'requestParams': request_data,
            'timestamp': datetime.now().isoformat(),
            'requestId': str(uuid.uuid4())
        }
        
        logger.info(f"请求处理成功，返回状态: success")
        return jsonify(response_data)
        
    except ValueError as val_error:
        logger.error(f"数据验证错误: {str(val_error)}")
        return jsonify({
            'status': 'error',
            'message': f'数据格式错误: {str(val_error)}',
            'requestParams': request_data if 'request_data' in locals() else {}
        }), 400
    except Exception as e:
        logger.error(f"库存调整请求处理失败: {str(e)}", exc_info=True)
        return jsonify({
            'status': 'error',
            'message': f'请求处理失败: {str(e)}',
            'requestParams': request_data if 'request_data' in locals() else {},
            'errorType': str(type(e).__name__)
        }), 500

# 获取预设参数接口
@inventory_adjustment_bp.route('/get-preset', methods=['GET'])
def get_preset():
    """获取预设参数"""
    try:
        logger.debug("开始处理获取预设参数请求")
        logger.debug(f"请求方法: {request.method}")
        logger.debug(f"请求参数: {request.args}")
        
        # 生成示例数据用于前端预览
        current_time = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        business_no = str(uuid.uuid4().fields[-1])[:13]
        source_code = f"{business_no}112"
        
        logger.debug(f"生成预设数据: business_no={business_no}, source_code={source_code}, current_time={current_time}")
        
        preset_data = {
            "appointmentNo": "",
            "businessNo": business_no,
            "checkMethod": "20",
            "detailList": [
                {
                    "actualArrivalNumber": 150000,
                    "batchCode": "1",
                    "lineNo": "1",
                    "planArrivalNumber": 150000,
                    "productDate": "2025-07-01",
                    "sampleQuality": 7,
                    "skuCode": "6926523473692",
                    "sourceCode": source_code,
                    "volume": 1,
                    "weight": 1
                }
            ],
            "forecastArrivalTime": current_time,
            "forecastDeliveryTime": current_time,
            "isCallCar": 0,
            "orderCreator": "LY",
            "remark": "LY",
            "sourceType": "WWJG",
            "supplierCode": "CO00049541",
            "totalVolume": 1,
            "totalWeight": 1,
            "warehouseCode": "DCN",
            "warehouseOutCode": "DCN"
        }
        
        logger.info("预设参数生成成功，准备返回数据")
        logger.debug(f"返回的预设数据: {json.dumps(preset_data, ensure_ascii=False)}")
        
        return jsonify({
            'status': 'success',
            'data': preset_data
        })
        
    except Exception as e:
        logger.error(f"获取预设参数失败: {str(e)}", exc_info=True)
        return jsonify({
            'status': 'error',
            'message': f'获取预设参数失败: {str(e)}',
            'errorType': str(type(e).__name__)
        }), 500

File: app/static/js/inventory_adjustment.js
/**
 * 库存调整页面JavaScript
 * 现代化的商务风格交互体验
 */

class InventoryAdjustmentManager {
    constructor() {
        this.form = document.getElementById('inventory-adjustment-form');
        this.skuCodeInput = document.getElementById('skuCode');
        this.quantityInput = document.getElementById('quantity');
        this.warehouseCodeInput = document.getElementById('warehouseCode');
        this.apiEnvSelect = document.getElementById('apiEnv');
        this.jsonPreview = document.getElementById('json-preview');
        this.resultCard = document.getElementById('result-card');
        this.resultStatus = document.getElementById('result-status');
        this.resultContent = document.getElementById('result-content');
        this.submitBtn = document.getElementById('submit-btn');
        this.previewBtn = document.getElementById('preview-btn');
        this.resetBtn = document.getElementById('reset-btn');
        this.copyJsonBtn = document.getElementById('copy-json-btn');
        
        this.init();
    }

    init() {
        this.bindEvents();
        this.loadPresetData();
    }

    bindEvents() {
        // 表单提交
        this.form.addEventListener('submit', (e) => this.handleSubmit(e));
        
        // 输入变化时更新预览
        [this.skuCodeInput, this.quantityInput, this.warehouseCodeInput, this.apiEnvSelect].forEach(element => {
            element.addEventListener('input', () => this.updateJsonPreview());
            element.addEventListener('change', () => this.updateJsonPreview());
        });
        
        // 预览按钮
        this.previewBtn.addEventListener('click', () => {
            this.updateJsonPreview();
            this.showToast('参数预览已更新', 'info');
        });
        
        // 重置按钮
        this.resetBtn.addEventListener('click', () => {
            this.form.reset();
            this.resultCard.style.display = 'none';
            this.jsonPreview.textContent = '{}';
            this.showToast('表单已重置', 'info');
        });
        
        // 复制JSON按钮
        this.copyJsonBtn.addEventListener('click', () => {
            const jsonText = this.jsonPreview.textContent;
            navigator.clipboard.writeText(jsonText).then(() => {
                this.showToast('参数已复制到剪贴板', 'success');
            }).catch(() => {
                this.showToast('复制失败，请手动复制', 'error');
            });
        });
    }

    loadPresetData() {
        // 加载预设数据用于预览
        console.log('开始加载预设数据...');
        fetch('/inventory_adjustment/get-preset')
            .then(response => {
                console.log('获取预设数据响应状态:', response.status);
                if (!response.ok) {
                    throw new Error(`HTTP错误! 状态: ${response.status}`);
                }
                return response.json();
            })
            .then(data => {
                console.log('获取预设数据成功:', data);
                if (data.status === 'success') {
                    this.jsonPreview.textContent = JSON.stringify(data.data, null, 2);
                } else {
                    console.error('预设数据状态错误:', data.status);
                    this.showToast('加载预设数据失败: ' + data.message, 'error');
                    // 如果没有获取到预设数据，使用默认示例数据
                    this.updateJsonPreview();
                }
            })
            .catch(error => {
                console.error('加载预设数据失败:', error);
                this.showToast('加载预设数据失败: ' + error.message, 'error');
                // 如果请求失败，使用默认示例数据
                this.updateJsonPreview();
            });
    }

    updateJsonPreview() {
        // 获取表单数据
        const skuCode = this.skuCodeInput.value.trim();
        const quantity = parseInt(this.quantityInput.value) || 0;
        const warehouseCode = this.warehouseCodeInput.value.trim();
        const apiEnv = this.apiEnvSelect.value;
        
        // 如果没有输入，使用示例数据
        const currentTime = new Date().toLocaleString('zh-CN', {
            year: 'numeric',
            month: '2-digit',
            day: '2-digit',
            hour: '2-digit',
            minute: '2-digit',
            second: '2-digit',
            hour12: false
        }).replace(/\//g, '-');
        
        const businessNo = Math.floor(Math.random() * 10000000000000).toString().padStart(13, '0');
        const sourceCode = `${businessNo}112`;
        
        // 构建预览参数
        const previewData = {
            "appointmentNo": "",
            "businessNo": skuCode ? businessNo : "6926523473692",
            "checkMethod": "20",
            "detailList": [
                {
                    "actualArrivalNumber": quantity || 150000,
                    "batchCode": "1",
                    "lineNo": "1",
                    "planArrivalNumber": quantity || 150000,
                    "productDate": "2025-07-01",
                    "sampleQuality": 7,
                    "skuCode": skuCode || "6926523473692",
                    "sourceCode": sourceCode,
                    "volume": 1,
                    "weight": 1
                }
            ],
            "forecastArrivalTime": currentTime,
            "forecastDeliveryTime": currentTime,
            "isCallCar": 0,
            "orderCreator": "LY",
            "remark": "LY",
            "sourceType": "WWJG",
            "supplierCode": "CO00049541",
            "totalVolume": 1,
            "totalWeight": 1,
            "warehouseCode": warehouseCode || "DCN",
            "warehouseOutCode": warehouseCode || "DCN"
        };
        
        // 更新预览
        this.jsonPreview.textContent = JSON.stringify(previewData, null, 2);
    }

    handleSubmit(e) {
        e.preventDefault();
        
        // 验证表单
        if (!this.form.checkValidity()) {
            e.stopPropagation();
            this.form.classList.add('was-validated');
            this.showToast('请填写所有必填字段', 'error');
            console.log('表单验证失败，存在必填字段未填写');
            return;
        }
        
        // 获取表单数据
        const formData = new FormData(this.form);
        const formValues = Object.fromEntries(formData.entries());
        
        // 显示加载状态
        this.submitBtn.disabled = true;
        this.submitBtn.innerHTML = '<i class="fas fa-spinner fa-spin me-1"></i>处理中...';
        
        // 提交请求
        console.log('开始提交库存调整请求，表单数据:', formValues);
        console.log('当前选择的API环境:', this.apiEnvSelect.value);
        
        fetch('/inventory_adjustment/submit', {
            method: 'POST',
            body: formData
        })
        .then(response => {
            console.log('接口响应状态码:', response.status);
            console.log('接口响应头信息:', response.headers);
            
            if (!response.ok) {
                throw new Error(`HTTP错误! 状态码: ${response.status}`);
            }
            
            return response.json().then(data => ({
                data: data,
                response: response
            }));
        })
        .then(({ data, response }) => {
            console.log('接口调用成功，完整响应数据:', data);
            
            // 显示结果
            this.resultCard.style.display = 'block';
            
            if (data.status === 'success') {
                this.resultStatus.className = 'badge bg-success';
                this.resultStatus.textContent = '成功';
                this.resultContent.textContent = JSON.stringify(data, null, 2);
                this.showToast(data.message || '请求成功', 'success');
                console.log('请求成功，返回消息:', data.message);
            } else {
                this.resultStatus.className = 'badge bg-danger';
                this.resultStatus.textContent = '失败';
                this.resultContent.textContent = JSON.stringify(data, null, 2);
                this.showToast(data.message || '请求失败', 'error');
                console.log('请求失败，返回状态:', data.status, '返回消息:', data.message);
            }
        })
        .catch(error => {
            console.error('接口调用失败，错误详情:', error);
            this.resultCard.style.display = 'block';
            this.resultStatus.className = 'badge bg-danger';
            this.resultStatus.textContent = '错误';
            this.resultContent.textContent = `请求处理失败: ${error.message}`;
            this.showToast('网络请求失败: ' + error.message, 'error');
        })
        .finally(() => {
            // 恢复按钮状态
            this.submitBtn.disabled = false;
            this.submitBtn.innerHTML = '<i class="fas fa-paper-plane me-1"></i>生成';
            console.log('请求处理完成，按钮状态已恢复');
        });
    }

    showToast(message, type = 'info') {
        // 创建toast元素
        const toast = document.createElement('div');
        toast.className = `toast align-items-center text-white ${this.getToastClass(type)} border-0 position-fixed bottom-5 right-5 z-50`;
        toast.role = 'alert';
        toast.setAttribute('aria-live', 'assertive');
        toast.setAttribute('aria-atomic', 'true');
        
        const iconClass = this.getToastIcon(type);
        
        toast.innerHTML = `
            <div class="d-flex">
                <div class="toast-body">
                    <i class="${iconClass} mr-2"></i>${message}
                </div>
                <button type="button" class="btn-close btn-close-white me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
            </div>
        `;
        
        // 添加到body
        document.body.appendChild(toast);
        
        // 显示toast
        const bootstrapToast = new bootstrap.Toast(toast, {
            autohide: true,
            delay: 3000
        });
        bootstrapToast.show();
        
        // 自动移除
        setTimeout(() => {
            toast.remove();
        }, 3500);
    }

    getToastClass(type) {
        const classes = {
            success: 'bg-success',
            error: 'bg-danger',
            warning: 'bg-warning',
            info: 'bg-info'
        };
        return classes[type] || classes.info;
    }

    getToastIcon(type) {
        const icons = {
            success: 'fas fa-check-circle',
            error: 'fas fa-exclamation-circle',
            warning: 'fas fa-exclamation-triangle',
            info: 'fas fa-info-circle'
        };
        return icons[type] || icons.info;
    }
}

// 页面加载完成后初始化
document.addEventListener('DOMContentLoaded', () => {
    new InventoryAdjustmentManager();
});